samza-core/src/main/java/org/apache/samza/processor/StreamProcessor.java

```diff
@@ -18,47 +18,178 @@
 import org.apache.samza.task.StreamTaskFactory;
 import org.apache.samza.util.Util;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import java.util.Map;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
 
 /**
- * StreamProcessor can be embedded in any application or executed in a distributed environment (aka cluster) as an
- * independent process.
  * <p>
- * <b>Usage Example:</b>
- * <pre>
- * StreamProcessor processor = new StreamProcessor(1, config);
- * processor.start();
- * try {
- *  boolean status = processor.awaitStart(TIMEOUT_MS);    // Optional - blocking call
- *  if (!status) {
- *    // Timed out
- *  }
- *  ...
- * } catch (InterruptedException ie) {
- *   ...
- * } finally {
- *   processor.stop();
- * }
- * </pre>
- * Note: A single JVM can create multiple StreamProcessor instances. It is safe to create StreamProcessor instances in
+ *   StreamProcessor can be embedded in any application or executed in a distributed environment (aka cluster) as an
+ *   independent process.
+ * </p>
+ *
+ * <b>Note</b>: A single JVM can create multiple StreamProcessor instances. It is safe to create StreamProcessor instances in
  * multiple threads.
  */
 @InterfaceStability.Evolving
 public class StreamProcessor {
+  private static final Logger LOGGER = LoggerFactory.getLogger(StreamProcessor.class);
+
   private final JobCoordinator jobCoordinator;
-  private final StreamProcessorLifecycleListener lifecycleListener;
-  private final String processorId;
+  private final StreamProcessorLifecycleListener processorListener;
+  private final Object taskFactory;
+  private final Map<String, MetricsReporter> customMetricsReporter;
+  private final Config config;
+  private final long taskShutdownMs;
+  private ExecutorService executorService;
+  private volatile SamzaContainer container = null;
+
+  @VisibleForTesting
+  JobCoordinatorListener jobCoordinatorListener = null;
+
+  // Latch used to synchronize between the JobCoordinator thread and the container thread, when the container is
+  // stopped due to re-balancing
+  private volatile CountDownLatch jcContainerLatch = new CountDownLatch(1);
+  private volatile boolean processorOnStartCalled = false;
+
+  SamzaContainer createSamzaContainer(ContainerModel containerModel, int maxChangelogStreamPartitions, JmxServer jmxServer) {
+    return SamzaContainer.apply(
+        containerModel,
+        config,
+        maxChangelogStreamPartitions,
+        jmxServer,
+        Util.<String, MetricsReporter>javaMapAsScalaMap(customMetricsReporter),
+        taskFactory);
+  }
+
+  JobCoordinatorListener createJobCoordinatorListener() {
+    return new JobCoordinatorListener() {
+
+      @Override
+      public void onJobModelExpired() {
+        if (container != null && container.getStatus().equals(SamzaContainerStatus.RUNNING)) {
```

Comment:
Just realized we still need the state check since calling shutdown on a stopped/failed container won't countdown the latch. 

I think we can change this check to `if (container != null && container != stopped or failed)`
That way if it's not started yet, it'll go to RUNNING -> STOPPED immediately. Does that sound reasonable?

As an aside, we should throw in SamzaContainer if run is called on a STOPPED | FAILED container.